### PR TITLE
*: replace gofail with the new failpoint implementation

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <chris@lonng.org>

### What problem does this PR solve? 

Replace [gofail](https://github.com/pingcap/gofail) with the new failpoint implementation [failpoint](https://github.com/pingcap/failpoint)

### What is changed and how it works?

Replace all dependencies and failpoints trigger condition.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

Related changes

### CODE REVIEW

**Comment by lonng** [comment]
@winkyao @bb7133 @crazycs520 @tiancaiamao @lysu  PTAL

File: ddl/index.go [hunk 1]
```diff
@@ -881,13 +883,14 @@ func (w *addIndexWorker) run(d *ddlCtx) {
...
 		logutil.Logger(ddlLogCtx).Debug("[ddl] add index worker got task", zap.Int("workerID", w.id), zap.String("task", task.String()))
-		// gofail: var mockAddIndexErr bool
-		//if w.id == 0 && mockAddIndexErr && !gofailMockAddindexErrOnceGuard {
-		//	gofailMockAddindexErrOnceGuard = true
-		//	result := &addIndexResult{addedCount: 0, nextHandle: 0, err: errors.Errorf("mock add index error")}
-		//	w.resultCh <- result
-		//	continue
-		//}
+		failpoint.Inject("mockAddIndexErr", func(val failpoint.Value) {
+			if w.id == 0 && val.(bool) && !gofailMockAddindexErrOnceGuard {
```
Comment (winkyao) [inline ddl/index.go]: this case used `gofailMockAddindexErrOnceGuard` to pass the concurrent tests, Could we use WithHook to replace it?

File: ddl/reorg.go [hunk 2]
```diff
@@ -375,12 +376,13 @@ func getReorgInfo(d *ddlCtx, t *meta.Meta, job *model.Job, tbl table.Table) (*re
...
 		logutil.Logger(ddlLogCtx).Info("[ddl] job get table range", zap.Int64("jobID", job.ID), zap.Int64("physicalTableID", pid), zap.Int64("startHandle", start), zap.Int64("endHandle", end))
 
-		// gofail: var errorUpdateReorgHandle bool
-		// if errorUpdateReorgHandle && !gofailOnceGuard {
-		//  // only return error once.
-		//	gofailOnceGuard = true
-		// 	return &info, errors.New("occur an error when update reorg handle.")
-		// }
+		failpoint.Inject("errorUpdateReorgHandle", func(val failpoint.Value) {
+			if val.(bool) && !gofailOnceGuard {
```
Comment (winkyao) [inline ddl/reorg.go]: use WithHook to replace `gofailOnceGuard `

**Comment by tiancaiamao** [comment]
LGTM

**Comment by siddontang** [comment]
@nolouch 

Please introduce to PD.

**Comment by siddontang** [comment]
@disksing 

do we need to use in go-client?

**Comment by disksing** [comment]
@siddontang No plan now. The current intend is based on mock-tikv to inject errors.

**Pull request merged**